### PR TITLE
ebi-ait/dcp-ingest-central#382: Populate ingest from valentine db

### PR DESCRIPTION
**Pull request by yusra-haider**
corresponding ticket: https://app.zenhub.com/workspaces/dcp-ingest-product-development-5f71ca62a3cb47326bdc1b5c/issues/ebi-ait/dcp-ingest-central/382

### CODE REVIEW

File: scripts/populate_ingest/convert/conversion_utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import re
+
+
+ACCESSION_PATTERNS = {
+    "insdc_project_accessions": "^[D|E|S]RP[0-9]+$",
```
Comment (amnonkhen) [inline scripts/populate_ingest/convert/conversion_utils.py:5]: The suffix `_accessions` is redundant.

File: scripts/populate_ingest/convert/conversion_utils.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    "ega_accessions": "EGA[DS][0-9]{11}",
+    "dbgap_accessions": "phs[0-9]{6}(\\.v[0-9])?(\\.p[0-9])?",
+    "geo_series_accessions": "^GSE.*$",
+    "array_express_accessions": "^E-....-.*$",
```
Comment (amnonkhen) [inline scripts/populate_ingest/convert/conversion_utils.py:9]: This pattern seems too broad.
After the 1st dash only letters are allowed.
After the 2nd one - only numbers.
See https://www.ebi.ac.uk/arrayexpress/help/accession_codes.html

File: scripts/populate_ingest/convert/conversion_utils.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    accessions = {}
+    for accession in data_accessions.split(','):
+        accession = accession.strip()
+        for key, pattern in ACCESSION_PATTERNS.items():
+            regex = re.compile(pattern)
+            if regex.match(accession):
+                accessions.setdefault(key, []).append(accession)
+    return accessions
+
+# ToDo: Move this to json-converter
```
Comment (amnonkhen) [inline scripts/populate_ingest/convert/conversion_utils.py:24]: Please don't leave todos in the code. Either do or delete.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+from convert.conversion_utils import get_accessions
+from convert.europe_pmc import EuropePmcConverter
+from data_entities.nxn_db import NxnDatabase
+from services.europe_pmc import EuropePmc
+from services.nxn_db import NxnDatabaseService
+from services.ingest import QuickIngest
+
+
+# INGEST_URL = "https://api.ingest.archive.data.humancellatlas.org"
+INGEST_URL = "http://localhost:8080"
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:23]: read config values from a config.py file that in turn reads them from environment variables. They way switching envs and running in a container would be easy.

See https://12factor.net/config

**Review by amnonkhen — CHANGES_REQUESTED** [review]

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        ingest_data_pub_urls = {pub.get("url") for pub in ingest_data_pubs if pub.get("url")}
+        ingest_data_pre_doi = {url.split('doi.org/')[1] for url in ingest_data_pub_urls if 'doi.org/' in url}
+
+        filter_doi = (self.nxn_data.get_values('DOI') | self.nxn_data.get_values('bioRxiv DOI')) - (
+                    ingest_data_pub_doi | ingest_data_pre_doi)
+        self.nxn_data.data = [row for row in self.nxn_data.data if self.nxn_data.get_value(row, 'DOI') in filter_doi or
+                              self.nxn_data.get_value(row, 'bioRxiv DOI') in filter_doi]
+
+    def __compare_on_accession__(self):
+        ingest_data_accessions = set(itertools.chain.from_iterable(
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:77]: This seems very repetitive. Consider adding the IDs to a list and mapping the elements.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            [data.get("biostudies_accessions") for data in self.ingest_data if data.get("biostudies_accessions")] +
+            [data.get("insdc_study_accessions") for data in self.ingest_data if data.get("insdc_study_accessions")]))
+
+        filter_accessions = self.nxn_data.get_values('Data location') - ingest_data_accessions
+        self.nxn_data.data = [row for row in self.nxn_data.data if self.nxn_data.get_value(row, 'Data location') in filter_accessions
+                              or not self.nxn_data.get_value(row, 'Data location')]
+
+    def __compare_on_title__(self):
+        ingest_data_pubs = list(itertools.chain.from_iterable([data.get("publications") for data in self.ingest_data if data.get("publications")]))
+        ingest_data_titles = set([reformat_title(pub.get("title")) for pub in ingest_data_pubs if pub.get("title")])
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:94]: Please use consistent quoting.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        self.nxn_data.data = [row for row in self.nxn_data.data if reformat_title(self.nxn_data.get_value(row, 'Title')) in filter_titles]
+
+    def compare(self):
+        self.__compare_on_doi__()
+        self.__compare_on_accession__()
+        self.__compare_on_title__()
+
+    def filter(self):
+        self.nxn_data.data = [row for row in self.nxn_data.data if
+                         self.nxn_data.get_value(row, 'Organism').lower() in ["human", "human, mouse", "mouse, human"]]
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:108]: The lists here should be constants.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        self.nxn_data.data = [row for row in self.nxn_data.data if
+                         any([tech.strip() in ["chromium", "drop-seq", "dronc-seq", "smart-seq2", "smarter",
+                                               "smarter (C1)"] for tech in
+                              self.nxn_data.get_value(row, 'Technique').lower().split("&")])]
+        self.nxn_data.data = [row for row in self.nxn_data.data if self.nxn_data.get_value(row, 'Measurement').lower() == 'rna-seq']
+
+    def __convert__(self, nxn_data_row) -> dict:
+        ingest_project = {
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:116]: This could be refactored to be made of functions in appropriate level of abstraction:
- make_ingest_Project
- add_publication_info
- add_accesstion

and so on.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            'content': {
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:117]: Do we have any room to document that this project was added by this script, such as a comments field that would say "added by XYZ script" maybe including the date.

**Review by amnonkhen — CHANGES_REQUESTED** [review]

**Comment by amnonkhen** [comment]
This script needs some testing.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+                response = self.ingest_api.new_project(project)
+                uuid = response.get('uuid', {}).get('uuid')
+                added_projects.append(uuid)
+                logging.info(f'added to ingest with uuid {uuid}')
+        return added_projects
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Add projects to ingest from Valentine's / NXN database")
+    parser.add_argument("-u", "--url", type=str, help="Base URL for Ingest API", default=INGEST_URL)
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:158]: The url should have a default in the code, overridable by an env variable overridable by commandline.
It is easier and more common for tools such as docker, kubernetes, gitlab etc. to pass environmental information as environment variables.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    max_len = max(len(title1), len(title2))
+    dist_metric = 100-(Levenshtein.distance(title1, title2)/max_len)*100
+    return dist_metric
+
+def get_file_content(file_path):
+    with open(file_path, "r") as file_path:
+        return file_path.read()
+
+def prepare_logging():
+    logging.basicConfig(filename='nxn_db.log', format='%(asctime)s - %(levelname)s - %(message)s', level=logging.INFO)
```
Comment (aaclan-ebi) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:41]: Ah, it's nice you're saving logs into a file.  I usually just create a report file in the end.  I think it's a good practice for ad-hoc/ops scripts like this.

File: scripts/populate_ingest/populate_ingest_from_nxn.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        ingest_data_accessions = set(itertools.chain.from_iterable(
+            [data.get("insdc_project_accessions") for data in self.ingest_data if
+             data.get("insdc_project_accessions")] +
+            [data.get("geo_series_accessions") for data in self.ingest_data if data.get("geo_series_accessions")] +
+            [data.get("ega_accessions") for data in self.ingest_data if data.get("ega_accessions")] +
+            [data.get("dbgap_accessions") for data in self.ingest_data if data.get("dbgap_accessions")] +
+            [data.get("array_express_accessions") for data in self.ingest_data if
+             data.get("array_express_accessions")] +
+            [data.get("biostudies_accessions") for data in self.ingest_data if data.get("biostudies_accessions")] +
+            [data.get("insdc_study_accessions") for data in self.ingest_data if data.get("insdc_study_accessions")]))
```
Comment (aaclan-ebi) [inline scripts/populate_ingest/populate_ingest_from_nxn.py:86]: could just have 1 for loop

File: scripts/populate_ingest/services/ingest.py [hunk 13]
```diff
@@ -71,3 +71,9 @@ def project_exists(self, uuid: str) -> bool:
         results = response.json()
         count = results.get('page', {}).get('totalElements', 0)
         return count > 0
+
+    def get_all_projects(self):
```
Comment (aaclan-ebi) [inline scripts/populate_ingest/services/ingest.py:75]: There's a method call `get_all` from ingest_api.py . Do note that it returns an iterator and not a list.

**Review by aaclan-ebi — COMMENTED** [review]
Good job reusing populate script! Would be nice to have some unit tests esp for those that have complicated/specific logic.

File: scripts/populate_ingest/services/ingest.py [hunk 14]
```diff
@@ -71,3 +71,9 @@ def project_exists(self, uuid: str) -> bool:
...
+        # get ingest data as json
+        url = f'{self.url}/projects/?page=0&size=1000'
```
Comment (ESapenaVentura) [inline scripts/populate_ingest/services/ingest.py:77]: I see 2 issues here:
- We are using an unprotected endpoint (Which shouldn't be unprotected), so when we finally fix this issue the script will be unusable
- No pagination is added, so when we surpass the 1000 projects this will stop showing new projects

On the pagination issue, I think we could solve it by adding a small function:

```
def yield_all_pages(self, response, size):
    response_json = response.json()
    num_pages = response_json.get('page', {}).get('totalPages', 0)
    url = f"{self.url}/projects/?size={size}"
    for i in range(num_pages):
        response = self.get(f"{url}?page={i}&size={size}")
        yield response.json()
```

Then the next lines would become:
```suggestion
        size = 200
        url = f'{self.url}/projects/'
        response = self.get(url)
        projects = []
        for page in yield_all_pages(response, size):
            projects.extend(page.get('_embedded', {}).get('projects', [])
        return projects
```
I suggested a size but we could keep it as 1000. Anyways, there's still the issue of the authentication

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 9 above)
Comment (ESapenaVentura) [reply scripts/populate_ingest/populate_ingest_from_nxn.py:117]: That would probably fit in the "wranglingNotes" field

**Review by ESapenaVentura — CHANGES_REQUESTED** [review]
A couple of comments but other than that, great job!

File: scripts/populate_ingest/convert/conversion_utils.py (hunk 3 above)
Comment (yusra-haider) [reply scripts/populate_ingest/convert/conversion_utils.py:24]: This to-do was already there as part of the code, before I added things to it.

I'm unaware of the context of this comment, so I can't do or delete it.

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 4 above)
Comment (yusra-haider) [reply scripts/populate_ingest/populate_ingest_from_nxn.py:23]: this was changed in the latest commit, which specifies the default ingest url as the prod url and can take a url for some other environment as a script parameter.

Is that alright or should we go down the config route? there's only one constant so using a config feels a like an overkill

Comment (amnonkhen) [reply scripts/populate_ingest/populate_ingest_from_nxn.py:23]: It is a good practice to have configs in a dedicated file with the possibility of overriding by environment variables. If you think a single configuration parameter does not constitute a config file, so be it. It can be created for the next configuration param.
Just make sure you allow overriding using environment variables.

File: scripts/populate_ingest/convert/conversion_utils.py (hunk 1 above)
Comment (amnonkhen) [reply scripts/populate_ingest/convert/conversion_utils.py:5]: Thanks for clarifying.

File: scripts/populate_ingest/convert/conversion_utils.py (hunk 3 above)
Comment (amnonkhen) [reply scripts/populate_ingest/convert/conversion_utils.py:24]: If you don't know what it is, and cannot contact the person who wrote this todo, delete it.
It should not have been there in the first place.

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 9 above)
Comment (amnonkhen) [reply scripts/populate_ingest/populate_ingest_from_nxn.py:118]: Thanks @ESapenaVentura . Sounds like a good place.

Comment (yusra-haider) [reply scripts/populate_ingest/populate_ingest_from_nxn.py:118]: it's already part of wrangling notes: https://github.com/ebi-ait/hca-ebi-dev-team/blob/b4f914be8d00f3ae2a2cb84cd8698663a28069b0/scripts/populate_ingest/populate_ingest_from_nxn.py#L126

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 11 above)
Comment (amnonkhen) [reply scripts/populate_ingest/populate_ingest_from_nxn.py:43]: Logging could involve multiple locations in different granularities.
For example: high level events (start, stop, etc.) to the console, while deeper debug info goes in the log file.
see, for example: https://docs.python.org/3/howto/logging-cookbook.html#logging-to-multiple-destinations

File: scripts/populate_ingest/services/ingest.py (hunk 13 above)
Comment (yusra-haider) [reply scripts/populate_ingest/services/ingest.py:75]: would you recommend using that and removing `get_all_projects`?

Comment (aaclan-ebi) [reply scripts/populate_ingest/services/ingest.py:75]: yeah, i would prefer using `get_all`.

Comment (ESapenaVentura) [reply scripts/populate_ingest/services/ingest.py:75]: +1, disregard my comments on that. forgot about that!

File: scripts/populate_ingest/services/ingest.py (hunk 14 above)
Comment (ESapenaVentura) [reply scripts/populate_ingest/services/ingest.py:77]: Seeing @aaclan-ebi's comment above, disregard this pls :)

File: scripts/populate_ingest/convert/conversion_utils.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
+import re
+
+
+ACCESSION_PATTERNS = {
+    "insdc_project_accessions": "^[D|E|S]RP[0-9]+$",
+    "ega_accessions": "EGA[DS][0-9]{11}",
+    "dbgap_accessions": "phs[0-9]{6}(\\.v[0-9])?(\\.p[0-9])?",
+    "geo_series_accessions": "^GSE.*$",
```
[the remark this answers is not in the document]
Comment (ESapenaVentura) [reply scripts/populate_ingest/convert/conversion_utils.py:8]: We only wrangle datasets (Called "series" in GEO) which I think is fine

The rest of accessions are e.g. platforms (GPL...) 

Happy to be more restrictive than identifiers.org in that regard :)

File: scripts/populate_ingest/services/ingest.py (hunk 13 above)
Comment (aaclan-ebi) [reply scripts/populate_ingest/services/ingest.py:75]: ah yours has the same implementation as get_all i think :)

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 10 above)
Comment (amnonkhen) [reply scripts/populate_ingest/populate_ingest_from_nxn.py]: code would look like:

``` python
param = "default_value"
if env_var_exists:
   param = get_from_env()
if command_line_vlue_exists:
   param = get_from_command_line()
```

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 12 above)
Comment (yusra-haider) [reply scripts/populate_ingest/populate_ingest_from_nxn.py]: how's that?

File: scripts/populate_ingest/populate_ingest_from_nxn.py (hunk 11 above)
Comment (yusra-haider) [reply scripts/populate_ingest/populate_ingest_from_nxn.py]: @amnonkhen  I followed the cookbook above and that works, but now all exceptions get logged to the console as well.

I'd added the exception related code in this function: https://github.com/ebi-ait/hca-ebi-dev-team/blob/d3ad3f77f05710c3a35001ed313732891e686afb/scripts/populate_ingest/populate_ingest_from_nxn.py#L42 (following this stackoverflow answer: https://stackoverflow.com/questions/6234405/logging-uncaught-exceptions-in-python)
to make sure exceptions only get logged to the log file, but that's not working after the following the cookbook.

Would you have any ideas / suggestions?

File: .gitignore [hunk 16]
```diff
@@ -1,6 +1,8 @@
 .idea/
 venv/
+.venv/
 _local/
 docs/_site/
 docs/.jekyll-cache/
-.venv/
\ No newline at end of file
+.venv/
```
Comment (amnonkhen) [inline .gitignore:7]: Identical to line 3. Please remove one of them.

File: scripts/populate_ingest/requirements.txt [hunk 17]
```diff
@@ -6,4 +6,5 @@ argparse~=1.4.0
 json-converter
 assertpy
 mergedeep
-Levenshtein
\ No newline at end of file
+Levenshtein
```
Comment (amnonkhen) [inline scripts/populate_ingest/requirements.txt:9]: I would keep this file sorted.

**Comment by ESapenaVentura** [comment]
Summarised log from running the tool:

```
INFO - Running script to populate ingest with data from nxn db
INFO - Running program in dry run mode
INFO - using https://api.ingest.archive.data.humancellatlas.org/ for ingest API
INFO - GET schemas {"size": 20, "totalElements": 102, "totalPages": 6, "number": 1}
INFO - GET schemas {"size": 20, "totalElements": 102, "totalPages": 6, "number": 2}
INFO - GET schemas {"size": 20, "totalElements": 102, "totalPages": 6, "number": 3}
INFO - GET schemas {"size": 20, "totalElements": 102, "totalPages": 6, "number": 4}
INFO - GET schemas {"size": 20, "totalElements": 102, "totalPages": 6, "number": 5}
INFO - Finding and filtering entries in nxn db, to add to ingest
INFO - project count before filtering 1368
INFO - project count after filtering by organism 587
INFO - project count after filtering by technology 359
INFO - project count after filtering by measurement 357
INFO - found 357 nxn entries to add to ingest
INFO - Successfully added 0 project(s) to ingest
```

**Comment by ESapenaVentura** [comment]
After pulling for the latest changes (With a "projects_to_be_added.json" file), I have run a small check to see if the accessions referenced in the project were duplicated

I found that there are 2 duplications, but when looking into these projects in GEO, they seem to be pointing out to the same paper, so I think that is ok!

I am trying to think of tests that could be written, and given the nature of the script I can't think of many, but here are my 2 cents:

**Tests that should pass**
- Against an empty database, it should return a number of entries to be added
- Successfully write entries to ingest

**Tests that should not pass**
- Trying to write an entry that is already in ingest

I'll keep thinking, but I couldn't come up with much, sorry!

**tl;dr**: Run it, works as intended (in dry run), weird thing with an accession but solved issue. I thought about some tests

**Review by aaclan-ebi — APPROVED** [review]

File: scripts/populate_ingest/populate_ingest/convert/nxn_db.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+class NxnDatabaseConverter:
+    def __init__(self, ingest_schema, publication_service, publication_converter):
+        self.ingest_schema = ingest_schema
+        self.publication_service = publication_service
+        self.publication_converter = publication_converter
+
+    def convert(self, nxn_data):
+        converted_projects = []
+        for d in nxn_data.to_dict('records'):
```
Comment (amnonkhen) [inline scripts/populate_ingest/populate_ingest/convert/nxn_db.py:22]: Have you considered:

```python
converted_projects = map(lambda d: self.convert_row(d), nxn_data.to_dict('records'))
```

**Review by amnonkhen — CHANGES_REQUESTED** [review]

**Review by ESapenaVentura — APPROVED** [review]
LGTM!

**Review by amnonkhen — APPROVED** [review]
